File: server/entitlement/ports.ts
import type { VaultContext } from '../../lib/domain/identity';
import type { BillingApi } from '../billing/public';
import type { EntitlementProjectionRecord, OfflineLeaseRecord } from './core';
import type { EntitlementProjectionVersion, OfflineLeaseId } from './public';

export type EntitlementOwnershipPort = {
  owns(context: VaultContext): Promise<boolean>;
};

export type ProjectionCommitResult =
  | { readonly kind: 'applied' }
  | { readonly kind: 'conflict' };

export type OfflineLeaseCreateResult =
  | { readonly kind: 'issued' }
  | { readonly kind: 'replayed'; readonly lease: OfflineLeaseRecord }
  | { readonly kind: 'identifier-conflict' }
  | { readonly kind: 'projection-conflict' };

export type EntitlementRepository = {
  findProjection(
    context: VaultContext,
  ): Promise<EntitlementProjectionRecord | undefined>;
  commitProjection(input: {
    readonly expectedVersion: EntitlementProjectionVersion | undefined;
    readonly record: EntitlementProjectionRecord;
    readonly revokeActiveLeasesAt: number | null;
  }): Promise<ProjectionCommitResult>;
  findOfflineLease(
    context: VaultContext,
    leaseId: OfflineLeaseId,
  ): Promise<OfflineLeaseRecord | undefined>;
  createOfflineLease(input: {
    readonly expectedProjectionVersion: EntitlementProjectionVersion;
    readonly lease: OfflineLeaseRecord;
  }): Promise<OfflineLeaseCreateResult>;
};

export type EntitlementDependencies = {
  readonly billing: Pick<BillingApi, 'readSubscription'>;
  readonly ownership: EntitlementOwnershipPort;
  readonly repository: EntitlementRepository;
};

File: backend/internal/entitlement/ports.go
package entitlement

import (
	"context"

	"github.com/fukamu/notes/backend/internal/billing"
	"github.com/fukamu/notes/backend/internal/identity"
)

type BillingPort interface {
	ReadSubscription(context.Context, identity.VaultContext) (*billing.SubscriptionFacts, error)
}

type OwnershipPort interface {
	Owns(context.Context, identity.VaultContext) (bool, error)
}

type ProjectionCommitKind string

const (
	ProjectionApplied  ProjectionCommitKind = "applied"
	ProjectionConflict ProjectionCommitKind = "conflict"
)

type LeaseCreateKind string

const (
	LeaseCreateIssued             LeaseCreateKind = "issued"
	LeaseCreateReplayed           LeaseCreateKind = "replayed"
	LeaseCreateIdentifierConflict LeaseCreateKind = "identifier-conflict"
	LeaseCreateProjectionConflict LeaseCreateKind = "projection-conflict"
)

type LeaseCreateResult struct {
	Kind  LeaseCreateKind
	Lease *OfflineLeaseRecord
}

type Repository interface {
	FindProjection(context.Context, identity.VaultContext) (*ProjectionRecord, error)
	CommitProjection(
		context.Context,
		*ProjectionVersion,
		ProjectionRecord,
		*int64,
	) (ProjectionCommitKind, error)
	FindOfflineLease(context.Context, identity.VaultContext, OfflineLeaseID) (*OfflineLeaseRecord, error)
	CreateOfflineLease(context.Context, ProjectionVersion, OfflineLeaseRecord) (LeaseCreateResult, error)
}
